daemon/regmaild/regmaild.c

```diff
@@ -278,12 +272,206 @@ regmaildb_update_email(const char * userid, int userid_len, const char * email,
 end:
     if (Stmt != NULL)
         sqlite3_finalize(Stmt);
-    if (Db != NULL && !g_Db)
-        sqlite3_close(Db);
 
     return ret;
 }
 
+static void
+verifydb_count(const verifydb_req *req, verifydb_count_rep *rep)
+{
+    sqlite3 *Db = g_Db;
+    sqlite3_stmt *Stmt = NULL;
+    int r = SQLITE_OK;
+
+    rep->cb = sizeof(*rep);
+    rep->status = VERIFYDB_ERROR;
+
+    r = sqlite3_prepare(Db, "SELECT userid, generation FROM verifydb "
+            "WHERE vmethod = ? AND vkey = ?;", -1, &Stmt, NULL);
+    if (r != SQLITE_OK)
+        goto end;
+
+    r = sqlite3_bind_int64(Stmt, 1, req->vmethod);
+    if (r != SQLITE_OK)
+        goto end;
+    r = sqlite3_bind_text(Stmt, 2, req->vkey, -1, SQLITE_STATIC);
+    if (r != SQLITE_OK)
+        goto end;
+
+    while (sqlite3_step(Stmt) == SQLITE_ROW) {
+        userec_t u = {};
+        const char *userid = (const char *)sqlite3_column_text(Stmt, 0);
+        int64_t generation = sqlite3_column_int64(Stmt, 1);
+
+        if (userid == NULL)
+            continue;  // Bad record.
+
+        if (!passwd_load_user(userid, &u))
+            continue;  // No such user.
+
+        // Make sure the record corresponds to user of current generation.
+        if (u.firstlogin != generation) {
+            fprintf(stderr, "verifydb count: skipped old record: %s "
+                    "(generation: db %ld, current %d)\n",
+                    u.userid, generation, u.firstlogin);
+            continue;
+        }
+
+        if (strcasecmp(userid, req->userid) == 0)
+            rep->count_self++;
+        else
+            rep->count_other++;
+    }
+
+    rep->status = VERIFYDB_OK;
+
+end:
+    if (r != SQLITE_OK)
```

Comment:
I think this will become very hard to follow esp. in `verifydb_get`. As the performance isn't of concern here, `if` wouldn't hurt and I feel a single bail out point would be easier to follow.